ASCIIToSVG.php

```diff
@@ -304,10 +306,44 @@ public function popGroup() {
     $this->curGroup = array_pop($this->groupStack);
   }
 
+    /*
+    AAVZZ
+    Since SVG has no notion of layers, we emulate them by changing the order
+    of boxes. This involves tagging and reordering. Tagging is done in `addObject',
+    reordering in `rearrangeBoxes'. We tag a box with a layer prefix, like `a' (default),
+    `b', `c' and so on. So, the sequence numbers will be something like
+    a0, a1, b2, a3, a4, b5, c6 ...
+    Once all the boxes are parsed, we need to rearrange them like
+    a0, a1, a3, a4, b2, b5, c6 ...
+    and change the indices to be numeic (we rely on them being numeric!!)
+    XXX Some text lines belonging to overlying box are placed in underlying box
+    XXX and do not show up. Further investigation is necessary to place them correctly.
+    XXX This issue is not introduced with this patch.
+  */
+  
   public function addObject($o) {
-    $this->groups[$this->curGroup][] = $o;
+    if ($this->curGroup == "boxes") {
+      $layer = 'a';
+      if ($o->getOption('a2s:layer')) {
+          $layer = $o->getOption('a2s:layer');
+      }
+      $i =  $layer . $this->index++;
+      $this->groups[$this->curGroup][$i] = $o;
```

Comment:
The code above ends up calling `getOption` twice and uses `$this->curGroup` when it is already known to be 'boxes' based on the conditional. (That's another reason to remove the condition.) 

Another thing worth thinking about: if you do not layer all of your objects, you start to need to come up with creative names for layers since they are sorted lexicographically. If I want to stack my boxes below the default layer, I need to use uppercase names, some kinds of punctuation, or numbers for the start of the layer name. If I want to stack my boxes above the default layer, I _can't_ use those things for layer names.

I would recommend that we use a deterministic ordering, hence switching to `a2s:z-index`. Simply assign a number. Higher numbers mean closer to the front, lower numbers mean further to the back. This also removes the need to introduce the $index variable.

This can all be expressed in a single line of code which can replace most of this change.

```
$this->groups[$this->curGroup][$o->getOption('a2s:z-index') ?: 0];
```